tests/unit/kernels_reference.cpp

```diff
@@ -153,6 +153,147 @@ void dw::reference::SGDStep(Parameters& params, float learning_rate) {
     }
 }
 
+void dw::reference::CPUBatchNorm1DForward(const dw::Tensor& input, dw::Tensor& output,
+                                          dw::Tensor& input_centered, dw::Tensor& std,
+                                          dw::Tensor& running_mean, dw::Tensor& running_var,
+                                          bool is_training, float eps, float alpha,
+                                          const dw::Tensor& gamma, const dw::Tensor& beta) {
+    const auto& shape = input.shape();
+
+    int batch_size  = shape[0];
+    int in_features = shape[1];
+
+    float* raw_input          = input.data();
+    float* raw_output         = output.data();
+    float* raw_input_centered = input_centered.data();
+    float* raw_std            = std.data();
+    float* raw_running_mean   = running_mean.data();
+    float* raw_running_var    = running_var.data();
+    float* raw_gamma          = gamma.data();
+    float* raw_beta           = beta.data();
+
+    if (is_training) {
+        std::vector<float> input_mean(in_features);
```

Comment:
So, there are a lot of loops, a bit confused why it isn't a single loop ? here and bellow as well

Reply:
removed a few extra cycles